### PR TITLE
Fixes Map in IE

### PR DESCRIPTION
**Pull request by CheloXL**
This fixes the noted issue. 
Commit includes:
- Fix for Map not been able to use another map in ctor.
- Test for the above.
- Test for Set: pass another set to ctor and verify that the copy is correct. While this was working as expected, I think having a test for this is good as it is part of the spec.

### CODE REVIEW

**Comment by CheloXL** [comment]
Screenshoots...
map_fail
map_pass

File: polyfills/Map/polyfill.js [hunk 1]
```diff
@@ -43,11 +43,19 @@
...
 		this._values = [];
-
 		// If `data` is iterable (indicated by presence of a forEach method), pre-populate the map
-		data && (typeof data.forEach === 'function') && data.forEach(function (item) {
-			this.set.apply(this, item);
-		}, this);
+		if (data && (typeof data.forEach === 'function')){
+			if (data instanceof Map){
+				data.forEach(function (value, key) {
+					debugger;
```
Comment (JakeChampion) [inline polyfills/Map/polyfill.js]: rogue debugger statement, can this be removed please

File: polyfills/Map/polyfill.js [hunk 2]
```diff
@@ -43,11 +43,19 @@
 	var Map = function(data) {
 		this._keys = [];
 		this._values = [];
-
 		// If `data` is iterable (indicated by presence of a forEach method), pre-populate the map
-		data && (typeof data.forEach === 'function') && data.forEach(function (item) {
-			this.set.apply(this, item);
-		}, this);
+		if (data && (typeof data.forEach === 'function')){
+			if (data instanceof Map){
```
Comment (JakeChampion) [inline polyfills/Map/polyfill.js]: If you have a Map from an iframe or a worker or something, `Map` is different, and instanceof won't work. What do you think about assuming it is a `Map` if it has all the `Map` instance methods?

```
typeof Map.prototype.clear === 'function' &&
typeof Map.prototype.delete === 'function' &&
typeof Map.prototype.entries === 'function' &&
typeof Map.prototype.forEach === 'function' &&
typeof Map.prototype.get === 'function' &&
typeof Map.prototype.has === 'function' &&
typeof Map.prototype.keys === 'function' &&
typeof Map.prototype.set === 'function' &&
typeof Map.prototype.values === 'function'
```

This is very verbose, happy to discuss alternatives :-)

File: polyfills/Map/polyfill.js (hunk 1 above)
Comment (CheloXL) [reply polyfills/Map/polyfill.js]: Ups... will remove.

File: polyfills/Map/polyfill.js (hunk 2 above)
Comment (CheloXL) [reply polyfills/Map/polyfill.js]: Well... It might be verbose but should work. Will change

**Review by CheloXL — COMMENTED** [review]
@JakeChampion: Made changes based on your suggestions.
I left the instanceof check as a fastpath so I don't have to check for the existence of every method if I already know I have an instance of a Map.

File: polyfills/Element/prototype/matches/config.json [hunk 3]
```diff
@@ -10,6 +10,7 @@
 	],
 	"browsers": {
 		"ie": "6 - *",
+		"ie_mob": "10 - *",
```
Comment (JakeChampion) [inline polyfills/Element/prototype/matches/config.json]: @CheloXL Would you be able to post a screenshot of the tests passing on IE Mobile 10 & 11?

File: polyfills/Map/polyfill.js [hunk 4]
```diff
@@ -40,14 +40,37 @@
...
 		this._keys = [];
 		this._values = [];
-
 		// If `data` is iterable (indicated by presence of a forEach method), pre-populate the map
-		data && (typeof data.forEach === 'function') && data.forEach(function (item) {
-			this.set.apply(this, item);
-		}, this);
+		if (data && hasProtoMethod(data, 'forEach')){
+			// Fastpath: If `data` is a Map, shortcircuit all following the checks
+			if (data instanceof Map ||
```
Comment (JakeChampion) [inline polyfills/Map/polyfill.js:53]: 👍

File: polyfills/Element/prototype/matches/config.json (hunk 3 above)
Comment (CheloXL) [reply polyfills/Element/prototype/matches/config.json]: Sure... In any case, I think I still need to understand how pull requests works, as that was supposed to be ANOTHER pull request for another issue, but anyways..
<details><summary>Click here to view the screenshots</summary>


</details>

Comment (JakeChampion) [reply polyfills/Element/prototype/matches/config.json]: To make multiple PRs, you would need to create a separate branch for each PR.

@kentcdodds has made a great video tutorial on how to contribute to open source projects on Github. -- https://egghead.io/courses/how-to-contribute-to-an-open-source-project-on-github

Comment (CheloXL) [reply polyfills/Element/prototype/matches/config.json]: Thanks, reading right now. How do you want me to approach this? Should I create another branch (not sure how I can revert this commit here) for the above? Or can you simply kill two birds in one shoot?

Comment (JakeChampion) [reply polyfills/Element/prototype/matches/config.json]: I would prefer it to be two separate PRs if that is possible. To revert the commit you can do this:
```bash
git revert 2b50affe49cd7157bbf67a792480dc2b1ab5f777
```

If you are struggling to do this, I am happy to accept it as one PR :-)

**Review by JakeChampion — APPROVED** [review]

**Comment by triblondon** [comment]
Fixes Financial-Times/polyfill-service#1068

**Comment by triblondon** [comment]
@JakeChampion mergeable now?

**Pull request merged**